### PR TITLE
fix(regression): add created source files to spoon input

### PR DESCRIPTION
**Pull request by MartinWitt**
See (https://github.com/INRIA/spoon/issues/3366) for details. Currently there is only a test project with a simple javacc case and a addition to the CI as starting point for the bug.
Edit: this should now fully showcase the bug, with a minimal working example

### CODE REVIEW

**Comment by MartinWitt** [comment]
Looks fixed, but 
- [ ] Changelog still missing
- [ ] Remove hardcoded version
- [ ] Refactor the added source code
- [ ] Try fix for the xwiki bug
- [ ] Create a regression test from the xwiki snippet

Everything expect the last bulletpoint needs to be done before a merge.
@vmassol are there some deadlines you like to match?

File: src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java [hunk 1]
```diff
@@ -36,12 +36,26 @@ protected AbstractSpoonConfigurationBuilder(SpoonMojoGenerate spoon,
...
+			}
 		}
 		if(spoon.isIncludeTestDirectories()) {
 			srcDir.add(new File(spoon.getProject().getBuild().getTestSourceDirectory()));
+			if (!spoon.getSkipGeneratedSources()) {
+				spoon.getProject().addTestCompileSourceRoot(spoon.getProject().getBuild().getTestOutputDirectory());
+				for (String s : spoon.getProject().getTestCompileSourceRoots()) {
+					srcDir.add(new File(s));
+				}
+			}
```
Comment (MartinWitt) [inline src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:55]: looks awful and needs refactor

**Comment by vmassol** [comment]
@MartinWitt what I don't understand is that `getCompileSourceRoots()` should also return generated sources paths.

> @vmassol are there some deadlines you like to keep?

No it's ok, no urgency, we'll release XWiki 12.4 without this and it'll go in 12.5 (next month release).

Thanks a lot.

**Comment by MartinWitt** [comment]
My first thought was, that it should return it too. But looking https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html and https://stackoverflow.com/questions/30039613/maven-plugin-get-generated-sources-directory it seems like maven has no standard defined for this and cant resolve this for you. And I found no method to get generated-sources.

**Comment by vmassol** [comment]
Normally when you code a plugin that generates sources you add the target source directory to the source roots. Maybe the javacc plugin is not compliant and doesn't do it?

**Comment by vmassol** [comment]
> Maybe the javacc plugin is not compliant and doesn't do it?

ok I think that's the reason, the old codehaus javacc mojo is not respecting the rules from what I see: https://github.com/mojohaus/javacc-maven-plugin/blob/master/src/main/java/org/codehaus/mojo/javacc/JavaCCMojo.java#L170

**Comment by vmassol** [comment]
I see there's a new plugin version at https://github.com/phax/ph-javacc-maven-plugin

However, I don't see much changes:

https://github.com/phax/ph-javacc-maven-plugin/blob/master/src/main/java/org/codehaus/mojo/javacc/JavaCCMojo.java#L161

**Comment by MartinWitt** [comment]
Oh thank you for your research i have only searched the bug on our side. Now we change from a bug to supporting maven plugins that work "incorrectly", but we could fix this on our side. Javacc looks like a often used code generator, so dropping support is huge step.

**Comment by vmassol** [comment]
So I've upgraded to this new plugin and I have the same error on the spoon side.

**Comment by vmassol** [comment]
This is an example of a well-written plugin: https://github.com/antlr/antlr4/blob/master/antlr4-maven-plugin/src/main/java/org/antlr/mojo/antlr4/Antlr4Mojo.java#L307

As you can see it calls addTestCompileSourceRoots() at https://github.com/antlr/antlr4/blob/master/antlr4-maven-plugin/src/main/java/org/antlr/mojo/antlr4/Antlr4Mojo.java#L201

**Comment by vmassol** [comment]
Issue opened on the javacc plugin side at https://github.com/phax/ph-javacc-maven-plugin/issues/17

**Comment by MartinWitt** [comment]
You are really fast, thank you for your effort. Lets see what javacc developers say to your issue.
But we should keep in mind, that we could fix things like this on our site, if more maven plugins are not "good Maven citizen".

**Comment by vmassol** [comment]
Yes I don't think we should wait on them.

**Comment by MartinWitt** [comment]
hmm looks like my bad. You are 100% correct, it works without adding the lines.

**Comment by MartinWitt** [comment]
I changed the pom xwiki-rendering-wikimodel to force spoon 3.5-SNAPSHOT (version with fix) and rann mvn install -DskipTests > spoon.log. As you see in the log it works now?
```xml
      <plugin>
        <groupId>fr.inria.gforge.spoon</groupId>
        <artifactId>spoon-maven-plugin</artifactId>
        <version>3.5-SNAPSHOT</version>
      <executions>
        <execution>
          <goals>
            <goal>check</goal>
          </goals>
        </execution>
        </executions>
        <configuration>
          <includeTest>true</includeTest>
          <includeSource>true</includeSource>
          <noClasspath>false</noClasspath>
        </configuration>
        </plugin>```

[spoon.log](https://github.com/SpoonLabs/spoon-maven-plugin/files/4645876/spoon.log)

**Comment by vmassol** [comment]
>  As you see in the log it works now?

It does! Cool, thanks a lot :)

**Comment by MartinWitt** [comment]
The hardcoded versions can only be removed after release, because maven does not support use newest version like gradle. Only point left is create a real regression testcase for this. I hope I can get this done tomorrow/tonight.

**Comment by vmassol** [comment]
@MartinWitt  what hardcoded versions?

**Comment by MartinWitt** [comment]
https://github.com/SpoonLabs/spoon-maven-plugin/pull/70/files#diff-403631b4cad466f44faaca502d672427R38 like this.

File: src/test/projects/javacc-generate/pom.xml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            <groupId>net.java.dev.javacc</groupId>
+            <artifactId>javacc</artifactId>
+            <version>7.0.6</version>
+          </dependency>
+        </dependencies>
+      </plugin>
+    <plugin>
+      <groupId>fr.inria.gforge.spoon</groupId>
+      <artifactId>spoon-maven-plugin</artifactId>
+      <version>3.5-SNAPSHOT</version>
```
Comment (vmassol) [inline src/test/projects/javacc-generate/pom.xml:38]: Why don't you use a filtered resource for example?

See:
* https://github.com/xwiki/xwiki-commons/blob/46e0ef7112577803a00e0c7f1abc035be38a8bf2/xwiki-commons-tools/xwiki-commons-tool-xar/xwiki-commons-tool-xar-integration-tests/pom.xml#L55
* https://github.com/xwiki/xwiki-commons/blob/0a56fb5fb8c59ea93bc09a5897c929322b64265a/xwiki-commons-tools/xwiki-commons-tool-xar/xwiki-commons-tool-xar-integration-tests/src/test/resources/wrongVersion/pom.xml

Comment (MartinWitt) [reply src/test/projects/javacc-generate/pom.xml:38]: because i normally use gradle and didn't knew it. Going to have a look thx.

**Comment by MartinWitt** [comment]
To finish this bugfix/issue, the task were:

1.  Changelog still missing
2.  Remove hardcoded version
3.  Refactor the added source code
4.  Try fix for the xwiki bug
5.  Create a regression test from the xwiki snippet

1. is added, see [link](https://github.com/SpoonLabs/spoon-maven-plugin/pull/70/files#diff-04c6e90faac2675aa89e2176d2eec7d8R268-R274)
2. happens after merge, when we have a new version number.( @vmassol solution requires a resource file or i understood something wrong ( i'm far away from maven expert))
3. Done
4. See logfile, it works now.
5. I believe adding xwiki with generated files as a regression test only slows down the CI and creates a strange dependency if some errors occur. The current test should test the same in in smaller scale( use a generated java file in fullClassPath). 

Looks like all is done and @monperrus time for a review?

[https://github.com/INRIA/spoon/issues/3345](https://github.com/INRIA/spoon/issues/3345) is still open, but i believe a version like 3.4.1 shouldn't contain new features. This issue is part of patch 3.5?

File: src/test/projects/javacc-generate/pom.xml (hunk 2 above)
Comment (monperrus) [reply src/test/projects/javacc-generate/pom.xml:38]: What about simply removing ` <version>3.5-SNAPSHOT</version>` as you did above?

**Comment by monperrus** [comment]
Thanks a lot for the progress. LGTM, see last comment.

Do we already have a fix for https://github.com/INRIA/spoon/issues/3345?

If yes, let's merge it now and release 3.5 directly and save one release.
If no, let's release directly 3.4.1, per this PR.

File: src/test/projects/javacc-generate/pom.xml (hunk 2 above)
Comment (vmassol) [reply src/test/projects/javacc-generate/pom.xml:38]: If you remove the version the build won't be reproducible and maven will use the latest version available. You need a version (and you should actually use the maven enforcer to always make sure all versions are set).

**Comment by MartinWitt** [comment]
Looking [here](https://github.com/SpoonLabs/spoon-maven-plugin/blob/master/src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java#L63) we use fullclasspathmode by default. The issue is only changing a parameter?

**Comment by monperrus** [comment]
> Looking here we use fullclasspathmode by default.

So https://github.com/INRIA/spoon/issues/3345 is already fixed, and we can add it in the Changelog.

I think we can proceed, merge this PR and release 3.4.1. WDYT?

**Comment by MartinWitt** [comment]
just did a small readme update and yes i believe you can merge it.
@vmassol maybe a unrelated question but could we link to the xwiki usage, when you integrated it into it? A real world example is always nice to have.

**Comment by vmassol** [comment]
> @vmassol maybe a unrelated question but could we link to the xwiki usage, when you integrated it into it? A real world example is always nice to have.

Definitely, with pleasure :)

FYI, on the xwiki.org site we documented our spoon usages at https://dev.xwiki.org/xwiki/bin/view/Community/Building/#HAutomaticChecks

**Comment by monperrus** [comment]
Merged, thanks a lot @MartinWitt 

Now releasing 3.4.1

**Comment by monperrus** [comment]
FYI, 3.4.1 is in Maven Central

**Comment by vmassol** [comment]
Thanks @MartinWitt and @monperrus 

I've now tested v3.4.1 and I've found a regression: https://github.com/INRIA/spoon/issues/3376

I've rolled-back to v3.3 for now.

Thanks! (and sorry ;))

**Pull request merged**